storage/flashdrv/drv/grlib-spimctrl/flashdrv.c

```diff
@@ -0,0 +1,376 @@
+{
+	struct _storage_devCtx_t *ctx = strg->dev->ctx;
+	if (!common_isValidAddress(CFI_SIZE(ctx->cfi.chipSz), offs, len)) {
+		*retlen = 0;
+		return -EINVAL;
+	}
+
+	if (len == 0u) {
+		*retlen = 0;
+		return 0;
+	}
+
+	ssize_t ret = spimctrl_flash_readData(ctx, offs, buff, len);
+	if (ret < 0) {
+		*retlen = 0;
+		return ret;
+	}
+
+	*retlen = len;
+
+	return 0;
+}
+
+
+static ssize_t _flashdrv_mtdReadCb(uint64_t offs, void *buff, size_t len, cache_devCtx_t *ctx)
+{
+	storage_t *strg = ctx->strg;
+	size_t retlen;
+	ssize_t ret;
+
+	ret = _flashdrv_mtdRead(strg, offs, buff, len, &retlen);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return (ssize_t)retlen;
+}
+
+
+static int flashdrv_mtdRead(storage_t *strg, off_t offs, void *buff, size_t len, size_t *retlen)
+{
+	mutexLock(strg->dev->ctx->lock);
+	int ret = cache_read(strg->dev->ctx->cache, offs, buff, len);
+	mutexUnlock(strg->dev->ctx->lock);
+
+	if (ret < 0) {
+		*retlen = 0;
+	}
+	else {
+		*retlen = len;
+		ret = 0;
+	}
+
+	return ret;
+}
+
+
+static int _flashdrv_mtdWrite(storage_t *strg, off_t offs, const void *buff, size_t len, size_t *retlen)
+{
+	struct _storage_devCtx_t *ctx = strg->dev->ctx;
+	if (!common_isValidAddress(CFI_SIZE(ctx->cfi.chipSz), offs, len)) {
+		*retlen = 0;
+		return -EINVAL;
+	}
+
+	if (len == 0u) {
+		*retlen = 0;
+		return 0;
+	}
+
+	const uint8_t *src = buff;
+	size_t doneBytes = 0;
+
+	int res = 0;
+	const size_t pagesz = strg->dev->mtd->writeBuffsz;
+
+	while (doneBytes < len) {
+		size_t chunk = min(pagesz - (offs % pagesz), len - doneBytes);
+
+		res = spimctrl_flash_pageProgram(ctx, offs, src, chunk,
+				CFI_TIMEOUT_MAX_PROGRAM(ctx->cfi.toutTypical.bufWrite, ctx->cfi.toutMax.bufWrite));
+
+		if (res < 0) {
+			break;
+		}
+
+		doneBytes += chunk;
+		src += chunk;
+		offs += chunk;
+	}
+
+	*retlen = doneBytes;
+
+	return res;
+}
+
+
+static ssize_t _flashdrv_mtdWriteCb(uint64_t offs, const void *buff, size_t len, cache_devCtx_t *ctx)
+{
+	storage_t *strg = ctx->strg;
+	size_t retlen;
+	ssize_t ret;
+
+	if ((offs % strg->dev->mtd->writesz) != 0 || (len % strg->dev->mtd->writesz) != 0) {
+		return -EINVAL;
+	}
+
+	ret = _flashdrv_mtdWrite(strg, offs, buff, len, &retlen);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return (ssize_t)retlen;
+}
+
+
+static int flashdrv_mtdWrite(storage_t *strg, off_t offs, const void *buff, size_t len, size_t *retlen)
+{
+	mutexLock(strg->dev->ctx->lock);
+	int ret = cache_write(strg->dev->ctx->cache, offs, buff, len, LIBCACHE_POLICY);
+	mutexUnlock(strg->dev->ctx->lock);
+
+	if (ret < 0) {
+		*retlen = 0;
+	}
+	else {
+		*retlen = len;
+		ret = 0;
+	}
+
+	return ret;
+}
+
+
+static int flashdrv_mtdErase(storage_t *strg, off_t offs, size_t len)
+{
+	if ((strg == NULL) || (strg->dev == NULL) || (strg->dev->ctx == NULL)) {
+		return -EINVAL;
+	}
+
+	struct _storage_devCtx_t *ctx = strg->dev->ctx;
+	if (!common_isValidAddress(CFI_SIZE(ctx->cfi.chipSz), offs, len) || ((offs & (ctx->sectorsz - 1)) != 0) || (len % ctx->sectorsz != 0)) {
```

Comment:
Do we assume sectorsz = 2^n?
```suggestion
	if (!common_isValidAddress(CFI_SIZE(ctx->cfi.chipSz), offs, len) || ((offs & (ctx->sectorsz - 1)) != 0) || (len & (ctx->sectorsz - 1) != 0)) {
	
	or
	
	if (!common_isValidAddress(CFI_SIZE(ctx->cfi.chipSz), offs, len) || ((offs % ctx->sectorsz) != 0) || (len % ctx->sectorsz != 0)) {
```

Reply:
Probably shouldn't, gone with `%`